src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultStates = [];
+
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let stateCopy = { ...currentState };
+    const actionType = action.type;
+
+    if (actionType === 'addProperties') {
+      stateCopy = { ...stateCopy, ...action.extraData };
+    } else if (actionType === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete stateCopy[key];
+      }
+    } else if (actionType === 'clear') {
+      stateCopy = {};
+    }
```

Comment:
better to rewrite it using the switch case instead of multiple if/else statements